doc/m4.1
." DO NOT MODIFY THIS FILE!  It was generated by helptoman 1.1.1.1.
.TH M4 1 "October 1998" "GNU m4 1.4i" "GNU User's Manual"
.SH NAME
m4 \- short documentation for m4 1.4i
.SH SYNOPSIS
.B m4
[\fIOPTION\fR]...\fI \fR[\fIFILE\fR]...\fI\fR
.SH DESCRITION
.PP
Mandatory or optional arguments to long options are mandatory or optional
for short options too.
.SS "Operation modes:"
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.TP
\fB\-e\fR, \fB\-\-interactive\fR
unbuffer output, ignore interrupts
.TP
\fB\-E\fR, \fB\-\-fatal\-warnings\fR
stop execution after first warning
.TP
\fB\-Q\fR, \fB\-\-quiet\fR, \fB\-\-silent\fR
suppress some warnings for builtins
.TP
\fB\-P\fR, \fB\-\-prefix\-builtins\fR
force a `m4_' prefix to all builtins
.SS "Preprocessor features:"
.TP
\fB\-I\fR, \fB\-\-include\fR=\fIDIRECTORY\fR
search this directory second for includes
.TP
\fB\-D\fR, \fB\-\-define=NAME\fR[=\fIVALUE\fR]
enter NAME has having VALUE, or empty
.TP
\fB\-U\fR, \fB\-\-undefine\fR=\fINAME\fR
delete builtin NAME
.TP
\fB\-s\fR, \fB\-\-synclines\fR
generate `#line NO "FILE"' lines
.SS "Limits control:"
.TP
\fB\-G\fR, \fB\-\-traditional\fR
suppress all GNU extensions
.TP
\fB\-H\fR, \fB\-\-hashsize\fR=\fIPRIME\fR
set symbol lookup hash table size
.TP
\fB\-L\fR, \fB\-\-nesting\-limit\fR=\fINUMBER\fR
change artificial nesting limit
.SS "Frozen state files:"
.TP
\fB\-F\fR, \fB\-\-freeze\-state\fR=\fIFILE\fR
produce a frozen state on FILE at end
.TP
\fB\-R\fR, \fB\-\-reload\-state\fR=\fIFILE\fR
reload a frozen state from FILE at start
.SS "Debugging:"
.TP
\fB\-d\fR, \fB\-\-debug\fR=\fI[FLAGS]\fR
set debug level (no FLAGS implies `aeq')
.TP
\fB\-t\fR, \fB\-\-trace\fR=\fINAME\fR
trace NAME when it will be defined
.TP
\fB\-l\fR, \fB\-\-arglength\fR=\fINUM\fR
restrict macro tracing size
.TP
\fB\-o\fR, \fB\-\-error\-output\fR=\fIFILE\fR
redirect debug and trace output
.SS "FLAGS is any of:"
.TP
t
trace for all macro calls, not only traceon'ed
.TP
a
show actual arguments
.TP
e
show expansion
.TP
q
quote values as necessary, with a or e flag
.TP
c
show before collect, after collect and after call
.TP
x
add a unique macro call id, useful with c flag
.TP
f
say current input file name
.TP
l
say current input line number
.TP
p
show results of path searches
.TP
i
show changes in input files
.TP
V
shorthand for all of the above flags
.PP
If no FILE or if FILE is `-', standard input is read.
.SH SEE ALSO
The full documentation for
.B m4
is maintained as a Texinfo manual.  If the
.B info
and
.B m4
programs are properly installed at your site, the command
.IP
.B info m4
.PP
should allow you to access the manual as an hypertext.


NEWS
GNU M4 NEWS - User visible changes.

* Noteworthy changes in Version 1.6 (????-??-??) [stable]
  Released by ????, based on git versions 1.4.10b.x-* and 1.5.*

** Fix regressions introduced in 1.4.10b:
*** Using `builtin' or `indir' to perform nested `shift' calls triggered
    an assertion failure (not present in 1.4.11).
*** The command-line option -dV, as well as the builtin `debugmode(V)',
    failed to enable `t' and `c' debug options (not present in 1.4.11).
*** Comments that contain unbalanced quotes were not rescanned correctly
    when passed through $@ (not present in 1.4.11).
*** Using `defn' on a traced but undefined macro triggered an assertion
    failure (also present in 1.4.11, but not 1.4.12).

** Remove the undocumented command-line option '-N', as no one complained
   about the assertion failure regression that it introduced in 1.4.7.

** The `-o'/`--error-output' command-line options, which were replaced by
   `--debugfile' in 1.4.7, now issue a deprecation warning.  This warning
   harmlessly triggers with versions of Autoconf 2.60 and earlier, but can
   be silenced by applying this patch:
     http://git.sv.gnu.org/gitweb/?p=autoconf.git;a=commitdiff;h=714eeee87

** Fix the `m4wrap' builtin to accumulate wrapped text in FIFO order, as
   required by POSIX.  The manual mentions a way to restore the LIFO order
   present in earlier GNU M4 versions.  NOTE: this change exposes a bug
   in Autoconf 2.59 and earlier (which was fixed in Autoconf 2.60).

   If you want your package to work with pre-installed Autoconf without
   requiring 2.60, then add these lines to your project's configure.ac,
   prior to calling AC_INIT:

     # As long as this project is not ready to upgrade to autoconf 2.60
     # or newer, make sure that newer M4 will still use LIFO order:
     m4_define([m4_wrap], [m4_ifdef([_$0_text],
       [m4_define([_$0_text], [$1]m4_defn([_$0_text]))],
       [m4_define([_$0_text], [$1])m4_builtin([m4wrap],
         [m4_default(m4_defn([_$0_text])m4_undefine([_$0_text]))])])])

   On the other hand, if you want to install Autoconf 2.59 or earlier,
   then apply this patch:
     http://git.sv.gnu.org/gitweb/?p=autoconf.git;a=commitdiff;h=56d42fa71

** The `changecom' builtin semantics now match traditional
   implementations; if the start-comment string resembles a macro name or
   the start-quote string, comments are effectively disabled.

** The `divert' builtin now accepts an optional second argument of text
   that is immediately placed in the new diversion, regardless of whether
   the current expansion is nested within argument collection of another
   macro.  It has also been optimized for faster performance.

** The `substr' builtin now treats negative arguments as indices relative
   to the end of the string, and accepts an optional fourth argument of
   text to supply in place of the selected substring.  The manual gives an
   example of how to recover M4 1.4.x behavior, as well as an example of
   simulating the new negative argument semantics with older M4.

** The `index' builtin now takes an optional third argument as the index
   to begin searching from, with a negative argument relative to the end of
   the string.

** The `-d'/`--debug' command-line option now understands `-' and `+'
   modifiers, the way the builtin `debugmode' has always done; this allows
   `-d-V' to disable prior debug settings from the command line, similar to
   using the builtin `debugmode' without arguments.  The option
   `--debugmode' is added as an alias for `-d'.  The new flag `d' is added
   to control whether dereferencing an undefined macro causes a warning.
   The new flag `o' is added to control whether `dumpdef' outputs to stderr
   or the current `debugfile' location.  When the command line option is
   given the empty string, the mode is treated as `+adeq' instead of `aeq'.
   Also, the position of `-d' with respect to files on the command line is
   now significant.

** A new predefined text macro, `__m4_version__', expands to the unquoted
   version number of M4, if GNU extensions are enabled.  While you should
   generally favor feature tests over version number checks, this macro can
   be used, via `defn', to determine whether the version of m4 processing
   your file is adequate.

** The `defn', `popdef', and `undefine' builtins gained a new warning when
   operating on an undefined macro name, to match the warning already
   present in `builtin', `indir', and `dumpdef'.  For backwards
   compatibility, the warning can be disabled by using `debugmode(`-d')'
   (or the command line option `--debug=-d').  The flag is also cleared by
   the command line option `-E'/`--fatal-warnings', so that scripts written
   for 1.4.x do not cause the script to fail because of new warnings.

** Enhance the `indir' builtin to trace indirect macros, where the trace
   is requested via `traceon' or the command-line option `-t'.  Previously,

src/m4.c
/* GNU m4 -- A simple macro processor

   Copyright (C) 1989-1994, 2004-2014, 2016-2017, 2020-2025 Free
   Software Foundation, Inc.

   This file is part of GNU M4.

   GNU M4 is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   GNU M4 is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "m4.h"

#include <getopt.h>
#include <signal.h>
#include <stdarg.h>

#include "c-stack.h"
#include "progname.h"
#include "version-etc.h"

#define AUTHORS "Rene' Seindal", "Eric Blake"

/* Enable sync output for /lib/cpp (-s).  */
int sync_output = 0;

/* Debug (-d[flags]).  */
int debug_level = 0;

/* Disable GNU extensions (-G).  */
int no_gnu_extensions = 0;

/* Prefix all builtin functions by `m4_'.  */
int prefix_all_builtins = 0;

/* Max length of arguments in trace output (-lsize).  */
size_t max_debug_argument_length = SIZE_MAX;

/* Suppress warnings about missing arguments.  */
int suppress_warnings = 0;

/* If true, then warnings affect exit status.  */
static bool fatal_warnings = false;

/* If not zero, then value of exit status for warning diagnostics.  */
int warning_status = 0;

/* Artificial limit for expansion_level in macro.c.  */
int nesting_limit = 1024;

#ifdef ENABLE_CHANGEWORD
/* User provided regexp for describing m4 words.  */
const char *user_word_regexp = "";
#endif

/* Global catchall for any errors that should affect final error status, but
   where we try to continue execution in the meantime.  */
int retcode;

struct macro_definition
{
  struct macro_definition *next;
  int code;                     /* See label `defer'.  */
  const char *arg;
};
typedef struct macro_definition macro_definition;

/* Error handling functions.  */

/* Helper for all the error reporting, as a wrapper around
   error_at_line.  Report error message based on FORMAT and ARGS, on
   behalf of CALLER (if any), otherwise at the global current
   location.  If ERRNUM, decode the errno value that caused the error.
   If STATUS, exit immediately with that status.  If WARN, prepend
   'warning: '.  */
static void M4_GNUC_PRINTF (5, 0)
m4_verror_at_line (bool warn, int status, int errnum, const call_info *caller,
                   const char *format, va_list args)
{
  char *full = NULL;
  char *safe_macro = NULL;
  const char *macro = caller ? caller->name : NULL;
  size_t len = caller ? caller->name_len : 0;
  const char *file = caller ? caller->file : current_file;
  int line = caller ? caller->line : current_line;

  /* Sanitize MACRO, since we are turning around and using it in a
     format string.  The allocation is overly conservative, but
     problematic macro names only occur via indir or changeword.  */
  if (macro && memchr (macro, '%', len))
    {
      char *p = safe_macro = xcharalloc (2 * len);
      const char *end = macro + len;
      while (macro != end)
        {
          if (*macro == '%')
            {
              *p++ = '%';
              len++;
            }
          *p++ = *macro++;
        }
    }
  if (macro)
    /* Use slot 1, so that the rest of the code can use the simpler
       quotearg interface in slot 0.  */
    macro = quotearg_n_mem (1, safe_macro ? safe_macro : macro, len);
  /* Prepend warning and the macro name, as needed.  But if that fails
     for non-memory reasons (unlikely), then still use the original
     format.  */
  if (warn && macro)
    full = xasprintf (_("warning: %s: %s"), macro, format);
  else if (warn)
    full = xasprintf (_("warning: %s"), format);
  else if (macro)
    full = xasprintf (_("%s: %s"), macro, format);
  verror_at_line (status, errnum, line ? file : NULL, line,
                  full ? full : format, args);
  free (full);
  free (safe_macro);
  if ((!warn || fatal_warnings) && !retcode)
    retcode = EXIT_FAILURE;
}

/* Wrapper around error.  Report error message based on FORMAT and
   subsequent args, on behalf of CALLER (if any), and the current
   input line (if any).  If ERRNUM, decode the errno value that caused
   the error.  If STATUS, exit immediately with that status.  */
void
m4_error (int status, int errnum, const call_info *caller,
          const char *format, ...)
{
  va_list args;
  va_start (args, format);
  if (status == EXIT_SUCCESS && warning_status)
    status = EXIT_FAILURE;
  m4_verror_at_line (false, status, errnum, caller, format, args);
  va_end (args);
}

/* Wrapper around error.  Report warning message based on FORMAT and
   subsequent args, on behalf of CALLER (if any), and the current
   input line (if any).  If ERRNUM, decode the errno value that caused
   the warning.  */
void
m4_warn (int errnum, const call_info *caller, const char *format, ...)
{
  va_list args;
  if (!suppress_warnings)
    {
      va_start (args, format);
      m4_verror_at_line (true, warning_status, errnum, caller, format, args);
      va_end (args);
    }
}


/* Print a usage message and exit with STATUS.  */
static void M4_GNUC_NORETURN
usage (int status)
{
  if (status != EXIT_SUCCESS)
    xfprintf (stderr, _("Try `%s --help' for more information.\n"),
              program_name);
  else
    {
      xprintf (_("Usage: %s [OPTION]... [FILE]...\n"), program_name);
      fputs (_("\
Process macros in FILEs.  If no FILE or if FILE is `-', standard input\n\
is read.\n\
"), stdout);
      puts ("");
      fputs (_("\
Mandatory or optional arguments to long options are mandatory or optional\n\
for short options too.\n\
\n\
Operation modes:\n\
      --help                   display this help and exit\n\
      --version                output version information and exit\n\
"), stdout);
      xprintf (_("\
  -E, --fatal-warnings         once: warnings become errors, twice: stop\n\
                                 execution at first error\n\
  -i, --interactive            unbuffer output, ignore interrupts\n\
  -P, --prefix-builtins        force a `m4_' prefix to all builtins\n\
  -Q, --quiet, --silent        suppress some warnings for builtins\n\
      --warn-macro-sequence[=REGEXP]\n\
                               warn if macro definition matches REGEXP,\n\
                                 default %s\n\
"), DEFAULT_MACRO_SEQUENCE);
#ifdef ENABLE_CHANGEWORD
      fputs (_("\
  -W, --word-regexp=REGEXP     use REGEXP for macro name syntax\n\
"), stdout);
#endif
      puts ("");
      fputs (_("\
Preprocessor features:\n\
  -D, --define=NAME[=VALUE]    define NAME as having VALUE, or empty\n\
  -I, --include=DIRECTORY      append DIRECTORY to include path\n\
  -s, --synclines              generate `#line NUM \"FILE\"' lines\n\
  -U, --undefine=NAME          undefine NAME\n\
"), stdout);
      puts ("");
      xprintf (_("\
Limits control:\n\
  -g, --gnu                    override -G to re-enable GNU extensions\n\
  -G, --traditional            suppress all GNU extensions\n\
  -H, --hashsize=PRIME         set symbol lookup hash table size [509]\n\
  -L, --nesting-limit=NUMBER   change nesting limit, 0 for unlimited [%d]\n\
"), nesting_limit);
      puts ("");
      fputs (_("\
Frozen state files:\n\
  -F, --freeze-state=FILE      produce a frozen state on FILE at end\n\
  -R, --reload-state=FILE      reload a frozen state from FILE at start\n\
"), stdout);
      puts ("");
      fputs (_("\
Debugging:\n\
  -d, --debug[=[-|+]FLAGS], --debugmode[=[-|+]FLAGS]\n\
                               set debug level (no FLAGS implies `+adeq')\n\
      --debugfile[=FILE]       redirect debug and trace output to FILE\n\
                                 (default stderr, discard if empty string)\n\
  -l, --arglength=NUM          restrict macro tracing size\n\
  -t, --trace=NAME             trace NAME when it is defined\n\
"), stdout);
      puts ("");
      fputs (_("\
FLAGS is any of:\n\
  a   show actual arguments in trace\n\
  c   show collection line in trace\n\
  d   warn when dereferencing undefined macros\n\
  e   show expansion in trace\n\
  f   include current input file name in trace and debug\n\
  i   show changes in input files in debug\n\
  l   include current input line number in trace and debug\n\
"), stdout);
      fputs (_("\
  o   output dumpdef to stderr rather than debug file\n\
  p   show results of path searches in debug\n\
  q   quote values in dumpdef and trace, useful with a or e\n\
  t   trace all macro calls, regardless of per-macro traceon state\n\
  x   include unique macro call id in trace, useful with c\n\
  V   shorthand for all of the above flags\n\
"), stdout);
      puts ("");
      fputs (_("\
If defined, the environment variable `M4PATH' is a colon-separated list\n\
of directories included after any specified by `-I'.\n\
"), stdout);
      puts ("");
      fputs (_("\
Exit status is 0 for success, 1 for failure, 63 for frozen file version\n\
mismatch, or whatever value was passed to the m4exit macro.\n\
"), stdout);
      emit_bug_reporting_address ();
    }
  exit (status);
}

/* Decode options and launch execution.  */

/* For long options that have no equivalent short option, use a
   non-character as a pseudo short option, starting with CHAR_MAX + 1.  */
enum
{
  DEBUGFILE_OPTION = CHAR_MAX + 1,      /* no short opt */
  WARN_MACRO_SEQUENCE_OPTION,           /* no short opt */

  HELP_OPTION,                          /* no short opt */
  VERSION_OPTION                        /* no short opt */
};

static const struct option long_options[] =
{
  {"arglength", required_argument, NULL, 'l'},
  {"debug", optional_argument, NULL, 'd'},
  {"debugmode", optional_argument, NULL, 'd'},
  {"define", required_argument, NULL, 'D'},
  {"error-output", required_argument, NULL, 'o'}, /* FIXME: deprecate in 2.0 */
  {"fatal-warnings", no_argument, NULL, 'E'},
  {"freeze-state", required_argument, NULL, 'F'},
  {"gnu", no_argument, NULL, 'g'},
  {"hashsize", required_argument, NULL, 'H'},
  {"include", required_argument, NULL, 'I'},
  {"interactive", no_argument, NULL, 'i'},
  {"nesting-limit", required_argument, NULL, 'L'},
  {"prefix-builtins", no_argument, NULL, 'P'},
  {"quiet", no_argument, NULL, 'Q'},
  {"reload-state", required_argument, NULL, 'R'},
  {"silent", no_argument, NULL, 'Q'},
  {"synclines", no_argument, NULL, 's'},
  {"trace", required_argument, NULL, 't'},
  {"traditional", no_argument, NULL, 'G'},
  {"undefine", required_argument, NULL, 'U'},
  {"word-regexp", required_argument, NULL, 'W'},

  {"debugfile", optional_argument, NULL, DEBUGFILE_OPTION},
  {"warn-macro-sequence", optional_argument, NULL, WARN_MACRO_SEQUENCE_OPTION},

  {"help", no_argument, NULL, HELP_OPTION},
  {"version", no_argument, NULL, VERSION_OPTION},

  { NULL, 0, NULL, 0 },
};

/* Process a command line file NAME, and return true only if it was
   stdin.  */
static void
process_file (const char *name)
{
  if (STREQ (name, "-"))
    {
      /* If stdin is a terminal, we want to allow 'm4 - file -'
         to read input from stdin twice, like GNU cat.  Besides,
         there is no point closing stdin before wrapped text, to
         minimize bugs in syscmd called from wrapped text.  */
      /* TRANSLATORS: This is a short name for `standard input', used
         when a command line file was given as `-'.  */
      push_file (stdin, _("stdin"), false);
    }
  else
    {
      char *full_name;
      FILE *fp = m4_path_search (name, &full_name);
      if (fp == NULL)
        {
          error (0, errno, _("cannot open %s"),
                 quotearg_style (locale_quoting_style, name));
          /* Set the status to EXIT_FAILURE, even though we
             continue to process files after a missing file.  */
          retcode = EXIT_FAILURE;
          return;
        }
      push_file (fp, full_name, true);
      free (full_name);
    }
  expand_input ();
}

/* POSIX requires only -D, -U, and -s; and says that the first two
   must be recognized when interspersed with file names.  Traditional
   behavior also handles -s between files.  Starting OPTSTRING with
   '-' forces getopt_long to hand back file names as arguments to opt
   '\1', rather than reordering the command line.  */
#ifdef ENABLE_CHANGEWORD
#define OPTSTRING "-B:D:EF:GH:I:L:PQR:S:T:U:W:d::egil:o:st:"
#else
#define OPTSTRING "-B:D:EF:GH:I:L:PQR:S:T:U:d::egil:o:st:"
#endif

#ifdef DEBUG_REGEX
FILE *trace_file;
#endif /* DEBUG_REGEX */

int
main (int argc, char *const *argv, char *const *envp M4_GNUC_UNUSED)
{
  macro_definition *head;       /* head of deferred argument list */
  macro_definition *tail;
  macro_definition *defn;
  int optchar;                  /* option character */

  macro_definition *defines;
  bool interactive = false;
  bool seen_file = false;
  const char *debugfile = NULL;
  const char *frozen_file_to_read = NULL;
  const char *frozen_file_to_write = NULL;
  const char *macro_sequence = "";
  /* Hash table size (should be a prime) (-Hsize).  */
  size_t hash_table_size = HASHMAX;

  set_program_name (argv[0]);
  retcode = EXIT_SUCCESS;
  atexit (close_stdin);

#ifdef DEBUG_REGEX
  {
    const char *name = getenv ("M4_TRACE_FILE");
    if (name)
      trace_file = fopen (name, "a");
    if (trace_file)
      fputs ("m4:\n", trace_file);
  }
#endif /* DEBUG_REGEX */

  include_init ();
  debug_init ();
  set_quoting_style (NULL, escape_quoting_style);
  set_char_quoting (NULL, ':', 1);
  if (c_stack_action (NULL) == 0)
    nesting_limit = 0;

  /* First, we decode the arguments, to size up tables and stuff.  */

  head = tail = NULL;

  while ((optchar = getopt_long (argc, (char **) argv, OPTSTRING,
                                 long_options, NULL)) != -1)
    switch (optchar)
      {
      default:
        usage (EXIT_FAILURE);

      case 'B':
      case 'S':
      case 'T':
        /* Compatibility junk: options that other implementations
           support, but which we ignore as no-ops and don't list in
           --help.  */
        error (0, 0, _("warning: `m4 -%c' may be removed in a future release"),
               optchar);
        break;

      case '\1':
        seen_file = true;
        /* fall through */
      case 'D':
      case 'U':
      case 's':
      case 't':
      case DEBUGFILE_OPTION:
      defer:
        /* Arguments that cannot be handled until later are accumulated.  */

        defn = (macro_definition *) xmalloc (sizeof (macro_definition));
        defn->code = optchar;
        defn->arg = optarg;
        defn->next = NULL;

        if (head == NULL)
          head = defn;
        else
          tail->next = defn;
        tail = defn;

        break;

      case 'E':
        debug_decode ("-d", SIZE_MAX);
        if (!fatal_warnings)
          fatal_warnings = true;
        else
          warning_status = EXIT_FAILURE;
        break;

      case 'F':
        frozen_file_to_write = optarg;
        break;

      case 'G':
        no_gnu_extensions = 1;
        break;

      case 'H':
        {
          long tmp = strtol (optarg, NULL, 10);
          hash_table_size = tmp;
        }
        if (hash_table_size == 0)
          hash_table_size = HASHMAX;
        break;

      case 'I':
        add_include_directory (optarg);
        break;

      case 'L':
        {
          long tmp = strtol (optarg, NULL, 10);
          nesting_limit = tmp;
        }
        break;

      case 'P':
        prefix_all_builtins = 1;
        break;

      case 'Q':
        suppress_warnings = 1;
        break;

      case 'R':
        frozen_file_to_read = optarg;
        break;

#ifdef ENABLE_CHANGEWORD
      case 'W':
        user_word_regexp = optarg;
        break;
#endif

      case 'd':
        if (seen_file)
          goto defer;
        if (debug_decode (optarg, SIZE_MAX) < 0)
          error (0, 0, _("bad debug flags: %s"),
                 quotearg_style (locale_quoting_style, optarg));
        break;

      case 'e':
        error (0, 0, _("warning: `%s' is deprecated, use `%s' instead"),
                       "-e", "-i");
        /* fall through */
      case 'i':
        interactive = true;
        break;

      case 'g':
        no_gnu_extensions = 0;
        break;

      case 'l':
        {
          long tmp = strtol (optarg, NULL, 10);
          max_debug_argument_length = tmp <= 0 ? SIZE_MAX : (size_t) tmp;
        }
        break;

      case 'o':
        /* -o/--error-output are deprecated synonyms of --debugfile,
           so issue a warning.  Don't call debug_set_output here, as
           it has side effects.  */
        error (0, 0, _("warning: `%s' is deprecated, use `%s' instead"),
               optchar == 'o' ? "-o" : "--error-output", "--debugfile");
        debugfile = optarg;
        break;

      case WARN_MACRO_SEQUENCE_OPTION:
         /* Don't call set_macro_sequence here, as it can exit.
            --warn-macro-sequence sets optarg to NULL (which uses the
            default regexp); --warn-macro-sequence= sets optarg to ""
            (which disables these warnings).  */
        macro_sequence = optarg;
        break;

      case VERSION_OPTION:
        version_etc (stdout, PACKAGE, PACKAGE_NAME, VERSION, AUTHORS, NULL);
        exit (EXIT_SUCCESS);
        break;

      case HELP_OPTION:
        usage (EXIT_SUCCESS);
        break;
      }

  defines = head;

  /* Do the basic initializations.  */
  if (debugfile && !debug_set_output (NULL, debugfile))
    m4_error (0, errno, NULL, _("cannot set debug file %s"),
              quotearg_style (locale_quoting_style, debugfile));

  input_init ();
  output_init ();
  symtab_init (hash_table_size);
  set_macro_sequence (macro_sequence);
  include_env_init ();

  if (frozen_file_to_read)
    reload_frozen_state (frozen_file_to_read);
  else
    builtin_init ();

  /* Interactive mode means unbuffered output, and interrupts ignored.  */

  if (interactive)
    {
      signal (SIGINT, SIG_IGN);
      setbuf (stdout, NULL);
    }

  /* Handle deferred command line macro definitions.  Must come after
     initialization of the symbol table.  */

  while (defines != NULL)
    {
      macro_definition *next;
      symbol *sym;
      const char *arg = defines->arg;

      switch (defines->code)
        {
        case 'D':
          assert (arg);
          {
            const char *value = strchr (arg, '=');
            size_t len = value ? (size_t) (value - arg) : strlen (arg);
            define_user_macro (arg, len, value ? value + 1 : "",
                               value ? SIZE_MAX : 0, SYMBOL_INSERT);
          }
          break;

        case 'U':
          assert (arg);
          lookup_symbol (arg, strlen (arg), SYMBOL_DELETE);
          break;

        case 'd':
          if (debug_decode (arg, SIZE_MAX) < 0)
            error (0, 0, _("bad debug flags: %s"),
                   quotearg_style (locale_quoting_style, optarg));
          break;

        case 't':
          assert (arg);
          sym = lookup_symbol (arg, strlen (arg), SYMBOL_INSERT);
          SYMBOL_TRACED (sym) = true;
          break;

        case 's':
          sync_output = 1;
          break;

        case '\1':
          assert (arg);
          process_file (arg);
          break;

        case DEBUGFILE_OPTION:
          if (!debug_set_output (NULL, arg))
             m4_error (0, errno, NULL, _("cannot set debug file %s"),
                       quotearg_style (locale_quoting_style,
                                       arg ? arg : _("stderr")));
          break;

        default:
          assert (!"main");
          abort ();
        }

      next = defines->next;
      free (defines);
      defines = next;
    }

  /* Handle remaining input files.  Each file is pushed on the input,
     and the input read.  Wrapup text is handled separately later.  */

  if (optind == argc && !seen_file)
    process_file ("-");
  else
    for (; optind < argc; optind++)
      process_file (argv[optind]);

  /* Now handle wrapup text.  */

  while (pop_wrapup ())
    expand_input ();

  /* Change debug stream back to stderr, to force flushing the debug
     stream and detect any errors it might have encountered.  The
     three standard streams are closed by close_stdin.  */
  debug_set_output (NULL, NULL);

  if (frozen_file_to_write)
    produce_frozen_state (frozen_file_to_write);
  else
    {
      make_diversion (0);
      undivert_all ();
    }
  output_exit ();
#ifndef NDEBUG
  /* Only spend time freeing memory to help isolate leaks; if
     assertions are disabled, save the time and exit now.  */
  free_regex ();
  quotearg_free ();
  symtab_free ();
#endif /* NDEBUG */
#ifdef DEBUG_REGEX
  if (trace_file)
    fclose (trace_file);
#endif /* DEBUG_REGEX */
  exit (retcode);
}
